src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/OptionsBuilderConfigurationExtensions.cs

```diff
@@ -65,12 +65,25 @@ public static class OptionsBuilderConfigurationExtensions
 
             optionsBuilder.Configure<IConfiguration>((opts, config) =>
             {
-                IConfiguration section = string.Equals("", configSectionPath, StringComparison.OrdinalIgnoreCase)
-                    ? config
-                    : config.GetSection(configSectionPath);
+                IConfiguration section = GetConfigurationSectionOrRoot(config, configSectionPath);
                 section.Bind(opts, configureBinder);
             });
+            optionsBuilder.Services.AddSingleton<IOptionsChangeTokenSource<TOptions>>(serviceProvider =>
+            {
+                var config = serviceProvider.GetRequiredService<IConfiguration>();
```

Comment:
I just wonder if there is a better way we would not need to call GetRequiredService here.

cc: @HaoK

Reply:
Hmm, currenlty I just the base `ConfigurationChangeTokenSource` from the `IConfiguration`, but I guess I could create a concrete custom type that implements the right interface here in this library which then would use ctor-injection to obtain the `IConfiguration`.

But then the problem is to pass the correct captured state (the `sectionName` and the `optionsBuilder.Name`) into that, so the code here would just become more complicated without much benefit.

What is the problem with calling `GetRequiredService` here?

Reply:
>  I could create a concrete custom type that implements the right interface here in this library which then would use ctor-injection to obtain the IConfiguration.

That would be better if we can avoid this call to `GetRequiredService` by using a class that ctor injects IConfiguration instead.

> What is the problem with calling GetRequiredService here?

Using the `GetRequiredService` is a service locator (anti-)pattern that we want to avoid wherever we can.

Reply:
I replaced the `IOptionsChangeTokenSource` registration with the follwing:

https://github.com/dotnet/runtime/blob/dcfd0becaad76c32de67479d62364f31aa0abafa/src/libraries/Microsoft.Extensions.Options.ConfigurationExtensions/src/OptionsBuilderConfigurationExtensions.cs#L73

The effect of this is however: The `Name` property of the `IOptionsChangeTokenSource<TOptions>` will be `OptionsBuilder.DefaultName` and **NOT** match the name passed in by the builder calling `BindConfiguration`. The `IConfiguration` instance from which the token change source will get its reload token is also going to be the root `IConfiguration` and not the particular section from which the `TOptions` instance was actually bound. Do we support section-scoped reload tokens?

Both these two drawbacks are necessary to avoid the call to `GetRequiredService`. I could not figure out how to do it otherwise.

Reply:
> Do we support section-scoped reload tokens?

not sure. @HaoK what do you recommend?

Reply:
No reload is a root config concept I believe https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Configuration.Abstractions/src/IConfigurationRoot.cs#L16

Reply:
In that case this approach is probably fine